### ISSUE TITLE
DeviceClient unusable after closing connection

### ISSUE THREAD

**Issue by johannesegger** [Azure/azure-iot-sdk-csharp#93]
We're using v1.2.9 of the IoT SDK and we noticed a weird behavior when the connection drops during message receiving. Take a look at the following test:

```
[Fact]
public async Task CompleteMessageHangs()
{
    const string serviceConnectionString = "...";
    
    var registryManager = RegistryManager.CreateFromConnectionString(serviceConnectionString);
    var serviceClient = ServiceClient.CreateFromConnectionString(serviceConnectionString);

    var device = await registryManager.AddDeviceAsync(new Device($"TestDevice-{Guid.NewGuid()}"));
    await serviceClient.SendAsync(device.Id, new Microsoft.Azure.Devices.Message(new byte[] { 0x01, 0x02, 0x03 }));

    var hostName = Microsoft.Azure.Devices.IotHubConnectionStringBuilder.Create(serviceConnectionString).HostName;
    var newDeviceConnectionString = Microsoft.Azure.Devices.Client.IotHubConnectionStringBuilder
        .Create(
            hostName,
            new DeviceAuthenticationWithRegistrySymmetricKey(device.Id, device.Authentication.SymmetricKey.PrimaryKey)
        )
        .ToString();

    var deviceClient = DeviceClient.CreateFromConnectionString(newDeviceConnectionString, Microsoft.Azure.Devices.Client.TransportType.Mqtt);
    deviceClient.OperationTimeoutInMilliseconds = (uint)TimeSpan.FromMinutes(1).TotalMilliseconds;
    var message = await deviceClient.ReceiveAsync(TimeSpan.FromMinutes(1));
    try
    {
        // TODO drop internet connection

        // this will time out
        await deviceClient.SendEventAsync(new Microsoft.Azure.Devices.Client.Message(new byte[] { 0x03, 0x02, 0x01 }));
    }
    finally
    {
        // this hangs forever because `DeviceClient.OnConnectionClosed` set `OperationTimeoutInMilliseconds` to `0`
        // and seems to hang somewhere in `IDelegatingHandler.RecoverConnections`
        await deviceClient.CompleteAsync(message);
    }
}
```

So the flow is the following: We receive a message, we send a reply and then we complete the original message. When the internet connection drops after the message is received (i.e. we can't send the reply) `deviceClient.SendEventAsync` will throw some exception after some timeout. The problem then occurs in the finally block because the call to `deviceClient.CompleteMessage` hangs forever because when `DeviceClient.OnConnectionClosed` was called, it set `DeviceClient.OperationTimeoutInMilliseconds` to `0` and then waits forever to recover connections and therefore never sets it back to `DefaultOperationTimeoutInMilliseconds` (see https://github.com/Azure/azure-iot-sdk-csharp/blob/2f5c034c/device/Microsoft.Azure.Devices.Client/DeviceClient.cs#L920).

So I think setting it to `0` is wrong (because recovering hangs now) and resetting it to `DefaultOperationTimeoutInMilliseconds` is also wrong, because it's not what the user might have set.

**Comment by jasmineymlo** [comment]
@johannesegger, Thanks for your comment.  Ideally, the recovering flow should bail out after the operation timeout expires (DeviceClient.OperationTimeoutInMilliseconds) as well.  We have an outstanding task for connection status feedback which would let the app knows about disconnection.  Before we have the notification, the sdk has to keep trying to resume the connection.  Otherwise, app could see silent disconnection.  We are going to work on the error notification in May.  Once that is in place,  the 2 mentioned lines of OperationTimeoutInMilliseconds will be removed.  The comment above the line in code also explains the situation.  Thanks

**Comment by johannesegger** [comment]
Thanks for your response. I'm glad that this is going to be fixed. In the meantime I just always set `OperationTimeoutInMilliseconds` to a non-zero value before calling any `DeviceClient` method, that should prevent any hangs.

**Comment by jasmineymlo** [comment]
fixed in commit 57e1b292dba824658860508263291844a4874858
